File: cameraAppliance/stream_video.py
import cv2 as cv
import base64
import socketio
import os
from dotenv import load_dotenv

# importing unique device id
# a unique id would be hard coded into every device
# the device's id would also be stored in firestore db for authentication
from device_CONFIG import device_id

# access environment variables
load_dotenv()
server_url = os.getenv("SERVER_URL")

sio = socketio.Client()

# construct the connection URL with the device ID as a query param
connection_url = f'{server_url}/api/stream-video?deviceId={device_id}'

# connect to server
sio.connect(connection_url, namespaces=['/api/stream-video'])

# check if connect was a success
if sio.connected:
    print('Connection to the server was successful.')
else:
    print('Failed to connect to the server.')

@sio.event
def connect():
    print('Connected to the server')
    

@sio.event
def disconnect():
    print('Disconnected from the server')

def send_video_frame(frame):
    # encode frame as JPEG
    _, img_encoded = cv.imencode('.jpg', frame)

    # convert encoded frame to base64 for transmitting over websocket
    base64_frame = base64.b64encode(img_encoded).decode('utf-8')

    try:
        # emit the frame to the server
        sio.emit('videoFrame', base64_frame, namespace='/api/stream-video')

    except Exception as err:
        print('Error sending frame:', err)

File: backend/routes/streamVideo.js
const { Server } = require("socket.io");

const handleVideoStreaming = (io) => {
    /*
        handles the video streaming namespace
    */

    const streamVideoNamespace = io.of("/api/stream-video");

    // connection event for namespace
    streamVideoNamespace.on("connection", (socket) => {
        const { deviceId } = socket.handshake.query;
        console.log(`Device ${deviceId} connected`);

        // join device-specific room via decvice unique id
        socket.join(deviceId);

        // handle incoming video frames
        socket.on("videoFrame", (frameData) => {
            // broadcast frame to other devices in the same room
            // (client, and any user with the device's id)
            socket.to(deviceId).emit("videoFrame", frameData);
        });

        socket.on("disconnect", () => {
            console.log(`Device ${deviceId} disconnected`);
        });
    });
};

module.exports = { handleVideoStreaming };
